### ISSUE TITLE
Possible memory leak when uploading blobs

### ISSUE THREAD

**Issue by sgarcialaguna-mms** [FineUploader/fine-uploader#1711]
## Type of issue [REQUIRED]
Bug

## Uploader type [REQUIRED]
traditional

### Fine Uploader version [REQUIRED]
5.5.1

### Browsers where the bug is reproducible [REQUIRED]
Firefox, Chrome, IE

### Operating systems where the bug is reproducible [REQUIRED]
Windows 7/8/10

### Exact steps required to reproduce the issue [REQUIRED]
Drag some sizable files into the dropzone (say, 10 100MB files).

### All of your Fine Uploader initialization JavaScript code [REQUIRED]
```
var receiverAddress = '127.0.0.1'
var receiverPort = '8000'
var receiverUrl = 'http://' + receiverAddress + ':' + receiverPort + '/upload/'

var queuedFiles = []

var uploader = new qq.FineUploaderBasic({
    debug: true,
    request: {
        endpoint: receiverUrl
    },
    callbacks: {
        onComplete: function(id, name)
        {
            document.getElementById('messages').innerHTML = 'Received ' + name + '<br>' + queuedFiles.length + ' files to go.'
            processNextFile()
        }
    }
})

var dropzone = document.getElementById('dropzone')
dropzone.addEventListener('dragover', function (event)
{
    event.stopPropagation()
    event.preventDefault()

    event.dataTransfer.dropEffect = 'move'
})

dropzone.addEventListener('drop', function (event)
{
    event.stopPropagation()
    event.preventDefault()

    queuedFiles = queuedFiles.concat(Array.prototype.slice.call(event.dataTransfer.files))
    processNextFile()
})

function processNextFile()
{
    var nextFile = queuedFiles.shift()
    if (nextFile)
    {
        processFile(nextFile)
    }
}

function processFile(file)
{
    var reader = new FileReader()
    reader.onload = function ()
    {
        console.log('Processing file ' + file.name)
        var arrayBuffer = this.result
        var byteArray = new Uint8Array(arrayBuffer)
        // Manipulate the byteArray in some way...
        var blob = new window.Blob([byteArray])
        uploader.addFiles({blob: blob, name: file.name})
    }
    reader.readAsArrayBuffer(file)
}
```

### Detailed explanation of the problem [REQUIRED]
Our code reads in each file, manipulates it (elided in the example), then adds the blob to the fineuploader library. Browser memory usage steadily rises and stays high even after the upload is complete - in some cases, the browser / tab crashes outright. We found that adding `this._handler.expunge(id)` in the `onComplete` callback is an effective workaround.

**Comment by sgarcialaguna-mms** [comment]
I played around with the example a bit more. After the upload is complete, I can call `uploader.getFile(0)` (or any other index) and that's where I see the blob still being held.

**Comment by rnicholus** [comment]
Thanks for the example and details. I'll put this near the top of my list due to the severity of this bug.

**Comment by rnicholus** [comment]
Ah, I see. Of course, the `Blob` is held onto by Fine Uploader. This is required in order to ensure the `getFile` method returns, well, the file. For `Blob` objects that are simply references to files (such as those submitted directly via the `<input type="file">` button or dropped into the dropzone, this isn't an issue - those aren't held in memory, they're held on disk. But once you take one of those files and manipulate them client-side, the resulting `Blob` is held in memory by the browser.

This might be a tricky one to fix without introducing breaking changes in the behavior of the library. I can't simply change to code to dispose of `Blob` objects that are not also `File` objects (which means they are held in memory) after the upload - this would be a breaking change and I'd like to avoid a breaking change to fix a bug. Also, there might be instances where integrators actually want to get back the `Blob` when calling `getFile`.

I think a reasonable solution may be to offer a new API method, something like `expungeBlob`. You could call it in an `onComplete` callback handler (or whenever you want) to remove the `Blob` from Fine Uploader's control and make it eligible for GC. 

With this new API method, your initialization code would look something like this:

```js
var uploader = new qq.FineUploaderBasic({
    request: {
        endpoint: receiverUrl
    },
    callbacks: {
        onComplete: function(id, name, response) {
           if (response.success) {
              this.expungeBlob(id)
           }
        }
    }
})
```

...and this should hopefully help to alleviate the memory issues. Thoughts?

**Comment by sgarcialaguna-mms** [comment]
Sounds good to me. Like I said, I'm already using a similar workaround anyway with `this._handler.expunge(id)`, so a _documented_ way of doing that is all I was looking for.

**Comment by rnicholus** [comment]
Sounds good - I'll have to add this method then, since accessing an internal function directly is not supported and may break with any release without notice.

**Comment by rnicholus** [comment]
Implemented and published as 5.14.0-beta1. Please see the [release notes](https://github.com/FineUploader/fine-uploader/releases/tag/5.14.0-beta1) for final implementation details, as they differ a bit from what was discussed here.
